### PR TITLE
chore: optimise pages API in view mode

### PR DESCRIPTION
**Pull request by subrata71**
## Description
This PR addresses issue https://github.com/appsmithorg/appsmith/issues/35881 by optimizing the pages API specifically for view mode. The primary goal is to enhance the performance and efficiency of the API, ensuring a smoother and faster user experience when viewing pages. The changes involve refactoring existing methods, introducing new ones, and updating test cases to align with the new functionality.

TL;DR
It includes the following changes
- Addition of projection for the query that fetches pages by applicationId in view mode
- Removal of sorting workspaces and their resources in view mode

Fixes https://github.com/appsmithorg/appsmith/issues/35881

## Automation

/ok-to-test tags="@tag.All"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/10610633615>
> Commit: 696b7787c0f72cc756122ee18620b0e14bfc6f4c
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=10610633615&attempt=3" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.All`
> Spec:
> <hr>Thu, 29 Aug 2024 08:52:24 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
  - Introduced a method to retrieve new pages by application ID with customizable field inclusion.
  - Enhanced data retrieval capabilities with more granular control over returned fields.
  - Added project-specific data handling in page retrieval logic.

- **Bug Fixes**
  - Improved logic for handling view mode in page retrieval based on application mode.

- **Documentation**
  - Updated tests to reflect changes in published page visibility, ensuring accurate assertions.

- **Refactor**
  - Streamlined methods for creating application pages and improved control flow for data retrieval.

- **Tests**
  - Enhanced asynchronous behavior in test methods for better execution control.

### CODE REVIEW

**Comment by sagar-qa007** [comment]
@coderabbitai generate unit testing code for this file.

**Comment by subrata71** [comment]
@sagar-qa007  I believe any extra tests are not necessary for this PR. Updating the existing test cases to align with the new changes should be sufficient. Please let me know your thought on this.

**Comment by subrata71** [comment]
@coderabbitai Please generate a PR description

File: app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java [hunk 1]
```diff
@@ -291,24 +292,25 @@ public Mono<ApplicationPagesDTO> findApplicationPagesByBranchedApplicationIdAndV
...
+            Mono<UserData> markedRecentlyAccessedMono = userDataService
                     .updateLastUsedResourceAndWorkspaceList(
                             branchedApplication.getId(),
                             branchedApplication.getWorkspaceId(),
                             WorkspaceResourceContext.APPLICATIONS)
-                    .then();
+                    .name(getQualifiedSpanName(MARK_RECENTLY_ACCESSED_RESOURCES_PAGES, applicationMode))
+                    .tap(Micrometer.observation(observationRegistry));
+
+            return Mono.zip(markedRecentlyAccessedMono, getApplicationPagesDTOMono)
```
Comment (sneha122) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java:309]: @subrata71 Curious, with Mono.zip the monos will be executed in parallel right? but if you see the previous code we were using .then on markedRecentlyAccessedMono, which means we want markedRecentlyAccessedMono to complete first and then execute getApplicationPagesDTOMono, will this not change the behaviour?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomNewPageRepositoryCEImpl.java [hunk 2]
```diff
@@ -48,6 +48,16 @@ public Flux<NewPage> findByApplicationId(String applicationId, AclPermission acl
                 .all();
     }
 
+    @Override
+    public Flux<NewPage> findByApplicationId(
+            String applicationId, AclPermission aclPermission, List<String> includeFields) {
+        return queryBuilder()
+                .criteria(Bridge.equal(NewPage.Fields.applicationId, applicationId))
+                .permission(aclPermission)
+                .fields(includeFields)
```
Comment (sneha122) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomNewPageRepositoryCEImpl.java:57]: Isn't .fields deprecated?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java (hunk 1 above)
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java:309]: @sneha122 They don't have any dependency hence running in parallel in edit mode won't change any behaviour

File: app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java [hunk 3]
```diff
@@ -385,7 +387,7 @@ public ApplicationPagesDTO getApplicationPagesDTO(
         return applicationPagesDTO;
     }
 
-    private static @NotNull PageNameIdDTO getPageNameIdDTO(NewPage pageFromDb, String homePageId, boolean viewMode) {
+    private static PageNameIdDTO getPageNameIdDTO(NewPage pageFromDb, String homePageId, boolean viewMode) {
```
[the remark this answers is not in the document]
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java:390]: This is not a standard @NotNull annotation rather coming from jetbrains (IDE).

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomNewPageRepositoryCEImpl.java (hunk 2 above)
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomNewPageRepositoryCEImpl.java:57]: Yes, it's marked deprecated but I don't see enough guideline for such projection that combines fields at the document-level and the fields inside either in unpublished/published attribute.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java (hunk 1 above)
Comment (sneha122) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/newpages/base/NewPageServiceCEImpl.java:309]: Got it, Thanks for the clarification

File: app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomNewPageRepositoryCEImpl.java (hunk 2 above)
Comment (subrata71) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/repositories/ce/CustomNewPageRepositoryCEImpl.java:57]: @sneha122 
Adding projections through projection classes necessitates creating a custom class for each variant of the projection, and the query must return data in these specific types. I'm not a fan of introducing numerous projection classes for different combinations or creating multiple queries when the underlying logic remains unchanged.

I would prefer to defer this concern and address it in a subsequent PR if required.

**Review by sneha122 — APPROVED** [review]

**Pull request merged**